### ISSUE TITLE
VisIt C++ API Question | how to highlight mesh vertex/edges?

### ISSUE THREAD

**Issue by biagas** [visit-dav/live-customer-response#215]
We're trying to [combine `visit-users` email list with GitHub issues](https://github.com/visit-dav/live-customer-response/wiki/How-the-new-GitHub-visit-users-Email-Integration-Works).
When replying on `visit-users`, please reply **only** to emails with a *Subject*
line that includes `[visit-dav/live-customer-response]`.

---
Hi,
I am developing the Nuclear Energy Advanced Modeling and Simulation (NEAMS) Workbench which has VisIt integrated for mesh data visualization, etc. and need to enhance the integration with mesh vertex/edge highlighting as a function of the interaction with the user input.
  
Basically, assuming the appropriate metadata is in place I want to change selected vertex and/or edge color (e.g., from black to red) to visually convey the mesh subset of interest to the user. The goal is to accelerate users in their development iterations when needing to assign/debug mesh parameter as specified in their application-specific input.
 
I am new to VisIt C++ API so my question is: are there examples/tutorials or a convenient API call that is available to do this render highlight?  
 
Best regards,
--
Rob Lefebvre
Tel: (865) 241-0496;  Fax: (865) 574-3527

**Comment by biagas** [comment]
Hello Rob,

How is your integration communicating/driving VisIt? Is it through the SimV2 interface?

Are you talking about using VisIt's Mesh plot?
There isn't separate functionality to render vertices and edges in different colors, but you can change the overall mesh color by changing MeshAttributes for the plot.

Kathleen

**Comment by markcmiller86-visit** [comment]
Hi Kathleen,
I am not using SimV2 interface, at least that I am aware.

I have the VisIt UI embedded via the use of the following:

    VSProxy* proxy;
    VisViewer *viewer;
    QvisGUI* guiApp;

Where VisViewer is modelled after the class described at https://www.visitusers.org/index.php?title=Using_the_viewer_in_a_Qt_application

I don’t know if there is a class or means by which I can access a class capable of conducting the highlight. Any direction on this would be great.

Thank you,
--
Rob Lefebvre
Tel: (865) 241-0496;  Fax: (865) 574-3527

From: Kathleen Biagas via visit-users <visit-users@elist.ornl.gov>
Reply-To: visit-dav/live-customer-response <reply@reply.github.com>, VisIt software users community <visit-users@elist.ornl.gov>
Date: Wednesday, January 27, 2021 at 2:29 PM
To: visit-dav/live-customer-response <live-customer-response@noreply.github.com>
Cc: Kathleen Biagas <notifications@github.com>, Subscribed <subscribed@noreply.github.com>
Subject: [visit-users] [EXTERNAL] Re: [visit-dav/live-customer-response] VisIt C++ API Question | how to highlight mesh vertex/edges? (#215)


Hello Rob,

How is your integration communicating/driving VisIt? Is it through the SimV2 interface?

Are you talking about using VisIt's Mesh plot?
There isn't separate functionality to render vertices and edges in different colors, but you can change the overall mesh color by changing MeshAttributes for the plot.

Kathleen

**Comment by biagas** [comment]
Thanks for the clarification.  

Using the SampleVisApp, you could add  'Mesh' as one of the plot options, along with the necessary modifications to be able to change the plot type to Mesh.

You would also want to add a method for changing the mesh color, similar to the setNContours method (also adding signals/slots).

I haven't looked at this in a long while so I will figure out how to specify the colors and get back with more specifics.

Kathleen

**Comment by biagas** [comment]
I modified SampleVisApp to allow Mesh as a plot option, but could not get the setting of the mesh color to work.
I've entered a ticket:
https://github.com/visit-dav/visit/issues/5409

I am attaching my modifications to the App for your perusal.

[SimpleVisApp.tar.gz](https://github.com/visit-dav/live-customer-response/files/5889297/SimpleVisApp.tar.gz)

**Comment by biagas** [comment]
As a workaround, if you still want to use VisIt's Mesh plot, you can open VisIt's gui, open Mesh Plot Attributes, change Mesh color source to 'Custom' then change the Mesh color to your desired highlight color, click 'Make default', then Save your settings.
At least for the SampleVisApp, this allowed Mesh plot to be rendered with the color I chose instead of the default black.

**Comment by markcmiller86-visit** [comment]
Hi Kathleen,
This is helpful to know.

Your solution points to changing the color of the entire mesh, correct?

I’m interested in a method that allows setting the color for only a portion of the mesh, as little as a single vertex or edge. Is that feasible?

Thanks,
--
Rob Lefebvre
Tel: (865) 241-0496;  Fax: (865) 574-3527

From: Kathleen Biagas via visit-users <visit-users@elist.ornl.gov>
Reply-To: visit-dav/live-customer-response <reply@reply.github.com>, VisIt software users community <visit-users@elist.ornl.gov>
Date: Thursday, January 28, 2021 at 2:01 PM
To: visit-dav/live-customer-response <live-customer-response@noreply.github.com>
Cc: Kathleen Biagas <notifications@github.com>, Comment <comment@noreply.github.com>, markcmiller86-visit <visit-users@ornl.gov>
Subject: [visit-users] [EXTERNAL] Re: [visit-dav/live-customer-response] VisIt C++ API Question | how to highlight mesh vertex/edges? (#215)


As a workaround, if you still want to use VisIt's Mesh plot, you can open VisIt's gui, open Mesh Plot Attributes, change Mesh color source to 'Custom' then change the Mesh color to your desired highlight color, click 'Make default', then Save your settings.
At least for the SampleVisApp, this allowed Mesh plot to be rendered with the color I chose instead of the default black.

**Comment by biagas** [comment]
Correct, my solution involves the entire mesh.
Within VisIt proper, a highlighted Pick on a Node can create a highlight on the selected vertex.
See our manual on [Pick](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/gui_manual/Quantitative/Pick.html) for more information.  You would have to use the 'Pick highlights' options.
There is also a 'Query' option for performing Pick (as opposed to using the mouse).
I'm really not sure if any of this mechanism can be tied into an embedded app.

No options to highlight an edge exist that I know of.

**Comment by biagas** [comment]
Hello Rob,

I was thinking more about your needs.  The [OnionPeel operator](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/gui_manual/Operators/OperatorTypes/OnionPeel_operator.html) may help solve part of your problem.
If used with a Mesh plot, it can serve as a highlight for a the edges of a specific cell, and as many cells adjacent to it as you like (by changing the number of 'layers').
It doesn't really highlight vertices, but can be used to highlight the edges of the cells adjacent to a chosen vertex.
